src/main/java/de/tum/in/www1/artemis/service/exam/StudentExamService.java

```diff
@@ -486,9 +477,53 @@ public void setUpExerciseParticipationsAndSubmissions(StudentExam studentExam, L
         }
     }
 
+    /**
+     * Starts all the exercises of all the student exams of an exam
+     *
+     * @param examId exam to which the student exams belong
+     * @return number of generated Participations
+     */
+    public int startExercises(Long examId) {
+        var exam = examRepository.findWithStudentExamsExercisesById(examId).orElseThrow(() -> new EntityNotFoundException("Exam", examId));
+        var studentExams = exam.getStudentExams();
+        List<StudentParticipation> generatedParticipations = Collections.synchronizedList(new ArrayList<>());
+        executeInParallel(() -> studentExams.parallelStream().forEach(studentExam -> setUpExerciseParticipationsAndSubmissions(studentExam, generatedParticipations)));
+        return generatedParticipations.size();
+    }
+
+    private void executeInParallel(Runnable task) {
+        final int numberOfParallelThreads = 10;
```

Comment:
Why exactly 10?
You could also retrieve the amount of available cores and then e.g. use 10% of them or something like that

Reply:
This method was not implemented by me  [(see StudentExamSerive:420)](https://github.com/ls1intum/Artemis/blob/develop/src/main/java/de/tum/in/www1/artemis/service/exam/StudentExamService.java). 
The change may occur, as I implemented a new Method  `setUpExerciseParticipationsAndSubmissionsWithInitializationDate`, which reuses the implemented functionality of the old `setUpExerciseParticipationsAndSubmissions`. The new one `setUpExerciseParticipationsAndSubmissions` (line 519) calls the `[...]WithInitializationDate` without passing an initialization Date.

Reply:
We should avoid such magic numbers - the amount of parallel threads should not be defined in a private method (and depends on the used system) :thinking:

Reply:
Since you're already refactoring, I thought it might be a good idea to improve the copied code as well